File: 01_GOF23/DP_examples/Python/02_Structural/Adapter/adapter.py
from adaptee import  AudioPlayer, VideoPlayer
from target import Player, IPlayer

# 适配器类
class PlayerAdapter(IPlayer):
    def __init__(self):
        self.player = Player()
        self.audio_player = AudioPlayer()
        self.video_player = VideoPlayer()

    def play(self, file_name, typ):
        # 调用原始播放器
        self.player.play(file_name, typ)
        # 调用音频播放器
        self.audio_player.play_media(file_name, typ)
        # 调用视频播放器
        self.video_player.play_media(file_name, typ)

File: 01_GOF23/DP_examples/Rust/02_Structural/Adapter/adapter.rs
use crate::adaptee::{AudioPlayer, VideoPlayer, MediaType, IAdvancedMediaPlayer};
use crate::target::{IPlayer, Player};

// 适配器类
pub struct PlayerAdapter {
    player: Player,
    audio_player: AudioPlayer,
    video_player: VideoPlayer,
}

impl PlayerAdapter {
    pub fn new() -> Self {
        Self {
            player: Player,
            audio_player: AudioPlayer,
            video_player: VideoPlayer,
        }
    }
}

impl IPlayer for PlayerAdapter {
    fn play(&self, file_name: &str, typ: MediaType) {
        // 调用原始播放器
        self.player.play(file_name, typ);
        // 调用音频播放器
        self.audio_player.play_media(file_name, typ);
        // 调用视频播放器
        self.video_player.play_media(file_name, typ);
    }
}
